### PR TITLE
calculate ref counts earlier to prevent bad clean

### LINKED ISSUE #9137: MismatchedBankHash when loading a mainnet-beta snapshot
STR:
1. Using the v1.0.10 release, run `solana-validator -l ledger -n mainnet-beta.solana.com:8001 -o -`

The snapshot produces a bank hash mismatch and the validator terminates: 
```
[2020-03-28T03:38:09.912189893Z INFO  solana_ledger::snapshot_utils] Loaded bank for slot: 2242884
[2020-03-28T03:38:09.912278093Z INFO  solana_runtime::accounts_db] total_stores: 188, newest_slot: 2242884, oldest_slot: 0, max_slot: 0 (num=5), min_slot: 1816112 (num=1)
[2020-03-28T03:38:09.912427593Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=188i
[2020-03-28T03:38:09.922820695Z WARN  solana_runtime::accounts_db] mismatched bank hash for slot 2242884: t6z86G59dJovkMtQ4f3VLKb4RCTHW2fWzKy7GrYuRFF (calculated) != 24LJe47dJQAbavHKhDZ8bP4iT7f1BuqT7s1wzpPyNnfX (expected)
[2020-03-28T03:38:09.922856295Z WARN  solana_runtime::accounts] verify_bank_hash failed: MismatchedBankHash
thread 'main' panicked at 'Snapshot bank for slot 2242884 failed to verify', ledger/src/snapshot_utils.rs:460:9
```

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Storage ref counts can prevent an account from being cleaned, but that also affects if another item would be cleaned. We don't go back and update the storage counts of those other items.

#### Summary of Changes

Calculate ref counts in earlier loop which does re-calculate the storage counts.

Fixes #9137

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -775,14 +775,20 @@ impl AccountsDB {
...
+            let mut would_unref_count = 0;
             for (_slot, account_info) in account_infos {
                 if *store_counts.get(&account_info.store_id).unwrap() != 0 {
                     no_delete = true;
                     break;
+                } else {
+                    would_unref_count += 1;
                 }
             }
+            if would_unref_count != accounts_index.ref_count_from_storage(&pubkey) {
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs]: [nits] Basically `would_unref_count` is same as `account_infos.len()`, so I think we could simplify this further. (Well, the original code should have done so... by me #8724; In other words, no need to count in fact; but i was too slanted to the concept of ref-COUNT-ing... xD):

```rust
if account_infos.len() != accounts_index.ref_count_from_storage(&pubkey) {
    no_delete = true;
} else {
    for (_slot, account_info) in account_infos {
        if *store_counts.get(&account_info.store_id).unwrap() != 0 {
            no_delete = true;
            break;
        }
    }
}
```

**Comment by ryoqun** [comment]
if possible, we'd like to have a test for this although writing one would be cumbersome as this is a bit a corner case.

**Comment by ryoqun** [comment]
@mvines Good news! I think this fixes yesterday's snapshot error with highly-likely chance. it seems that @sakridge and I independently reached to the same root cause and fix for it, although @sakridge was a ahead than me. ;)

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (sakridge) [reply runtime/src/accounts_db.rs]: Yea, that logic is better, I'll change to that. Thanks!
